src/main/java/eu/tsystems/mms/tic/testerra/plugins/xray/synchronize/AbstractXrayResultsSynchronizer.java

```diff
@@ -221,22 +221,29 @@ private synchronized void flushSyncQueue() {
         boolean areNewTestsToImport = testSetSyncQueue.stream().anyMatch(xrayTestSetIssue -> {
             return xrayTestSetIssue.getTestKeys().stream().anyMatch(key -> key.contains(XrayUtils.PREFIX_NEW_ISSUE));
         });
-        if (areNewTestsToImport) {
-            xrayTestExecutionImport.getResultTestIssueImport().getSuccess().forEach(jiraIssueReference -> {
-                try {
-                    // Replace the temporary key with real Jira key from the result 'xrayTestExecutionImport'
-                    JiraIssue issue = xrayUtils.getIssue(jiraIssueReference.getKey());
-                    testSetSyncQueue.forEach(xrayTestSetIssue -> {
-                        Optional<String> findKey = xrayTestSetIssue.getTestKeys().stream().filter(key -> key.contains(issue.getSummary())).findFirst();
-                        if (findKey.isPresent()) {
-                            xrayTestSetIssue.getTestKeys().removeIf(key -> key.contains(issue.getSummary()));
-                            xrayTestSetIssue.getTestKeys().add(issue.getKey());
-                        }
-                    });
-                } catch (IOException e) {
-                    log().error(String.format("Unable to read %s", IssueType.TestExecution), e);
-                }
-            });
+        if (areNewTestsToImport && xrayTestExecutionImport.getResultTestIssueImport() != null) {
+            if (xrayTestExecutionImport.getResultTestIssueImport().getError().size() > 0) {
+                log().error("Error at syncing with Jira");
+                xrayTestExecutionImport.getResultTestIssueImport().getError()
+                        .forEach(error -> log().error("{} - {}: {}", error.getProject(), error.getSummary(), error.getMessages().toString()));
+                return;
+            } else if (xrayTestExecutionImport.getResultTestIssueImport().getSuccess().size() > 0) {
+                xrayTestExecutionImport.getResultTestIssueImport().getSuccess().forEach(jiraIssueReference -> {
+                    try {
+                        // Replace the temporary key with real Jira key from the result 'xrayTestExecutionImport'
+                        JiraIssue issue = xrayUtils.getIssue(jiraIssueReference.getKey());
+                        testSetSyncQueue.forEach(xrayTestSetIssue -> {
+                            Optional<String> findKey = xrayTestSetIssue.getTestKeys().stream().filter(key -> key.contains(issue.getSummary())).findFirst();
+                            if (findKey.isPresent()) {
+                                xrayTestSetIssue.getTestKeys().removeIf(key -> key.contains(issue.getSummary()));
```

Comment:
I think this is not Stream work. We first find the entries we want to remove, ignore where we found them, and then search them again to remove and finally put their replacements at the end of the list.

This should probably be done in a for-each with `listIterator`:
```java
for (final var iterator = xrayTestSetIssue.getTestKeys().listIterator(); iterator.hasNext(); ) {
  final var key = iterator.next();
  if (key.contains(issue.getSummary()) {
    iterator.set(issue.getKey());
  }
}
```

Reply:
Ah... that is nice.